### PR TITLE
fix: Add support for `~test Array~string~` back in Class

### PR DESCRIPTION
**Pull request by sidharthv96**
## :bookmark_tabs: Summary

Add back support for `~test Array~string~`

Context: https://github.com/mermaid-js/mermaid/pull/4534#discussion_r1314143825

## :straight_ruler: Design Decisions

We cannot remove that support as it would be a breaking change. 

Also, if `~test Int` is supported, `~test Array~string~` should be too.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.

### CODE REVIEW

**Comment by Yokozuna59** [comment]
I think the PR title is misleading. Originally, I thought we're adding `~` also as a possible keyword to class name. But it appears you're just re-add the ability to have `~` (package) as a visibility when we have generic class.

File: packages/mermaid/src/diagrams/common/common.ts [hunk 1]
```diff
@@ -208,21 +208,33 @@ export const parseGenericTypes = function (input: string): string {
   return output.join('');
 };
 
+export const countOccurrence = (string: string, substring: string): number => {
+  return Math.max(0, string.split(substring).length - 1);
```
Comment (aloisklink) [inline packages/mermaid/src/diagrams/common/common.ts:212]: Can `string.split().length` ever be `0`? My feeling is that it can never be less than `1`, so the following should be safe.

```suggestion
  return string.split(substring).length - 1;
```

File: packages/mermaid/src/diagrams/common/common.ts [hunk 2]
```diff
@@ -208,21 +208,33 @@ export const parseGenericTypes = function (input: string): string {
   return output.join('');
 };
 
+export const countOccurrence = (string: string, substring: string): number => {
```
Comment (aloisklink) [inline packages/mermaid/src/diagrams/common/common.ts:211]: If we're `export`-ing this function from the file, it would be helpful to have a TSDoc comment, even if it's super basic and a bit repetitive for such a simple function:

```suggestion
/**
 * Count the number of times the given substring appears in the string.
 *
 * @param string - The input string.
 * @param substring - The string to search for.
 * @returns The number of times the substring appears in the string.
 */
export const countOccurrence = (string: string, substring: string): number => {
```

At some point in the future, we could add an ESLint rule like `"jsdoc/require-jsdoc": ["warn", {"publicOnly":true`}]` to warn us about exported functions that aren't documented.

**Review by aloisklink — COMMENTED** [review]
It's awesome to fix a breaking change, but I'm struggling to see what the breaking change this PR fixes actually is.

What is `~test Array~string~`? Can you give an example of a Mermaid diagram that has it (maybe also showing a screenshot, if the diagram only worked in an very old version of Mermaid)?

It would also be helpful if you could point to when support for `~test Array~string~` was broken (ideally to a specific commit or PR).

**Comment by sidharthv96** [comment]
@aloisklink this PR was actually targeted at https://github.com/mermaid-js/mermaid/pull/4534

The test case was removed in that PR. 

Context: https://github.com/mermaid-js/mermaid/pull/4534#discussion_r1314143825

File: packages/mermaid/src/diagrams/common/common.ts (hunk 2 above)
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/common/common.ts:211]: I prefer functions that are self explanatory. Comments are useful for explaining complex steps and functions, but seems like an unnecessary addition for simple functions.

File: packages/mermaid/src/diagrams/common/common.ts (hunk 1 above)
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/common/common.ts:212]: It can, when the input string is empty. I added the max call while writing the unit test when the "" case was added.

File: packages/mermaid/src/diagrams/common/common.ts (hunk 2 above)
Comment (aloisklink) [reply packages/mermaid/src/diagrams/common/common.ts:211]: > Comments are useful for explaining complex steps and functions, but seems like an unnecessary addition for simple functions.

I agree! But, I do think if we're exporting a function out of a file, the bar for documenting it is a lot lower, since you can't easily see the source-code of the function (especially if you're minifying code).

I think the name/params for this function are self-explanatory enough that we don't need a TSDoc comment, and it'll only be necessary if we ever add an ESLint rule like `"jsdoc/require-jsdoc": ["warn", {"publicOnly":true}]`.

File: packages/mermaid/src/diagrams/common/common.ts (hunk 1 above)
Comment (aloisklink) [reply packages/mermaid/src/diagrams/common/common.ts:212]: > It can, when the input string is empty. I added the max call while writing the unit test when the "" case was added.

:facepalm: That's my bad for reviewing your PR too quickly! You're right. `"".split("").length` returns `0`.

There's even a big warning about it on [`String.prototype.split()` MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split):
image

Great call on testing this edge case :smile:

**Review by aloisklink — APPROVED** [review]
> @aloisklink this PR was actually targeted at https://github.com/mermaid-js/mermaid/pull/4534
> 
> The test case was removed in that PR.
>
> Context: https://github.com/mermaid-js/mermaid/pull/4534#discussion_r1314143825

Ahh, awesome! :+1: Great catch on getting it fixed while it was still in the `develop` branch.

Btw, is it worth labeling this PR with **Skip changelog**? Since this bug has never been released to users (it's just on the `develop` branch), there's less of a need to put it in the Release notes (maybe current users of `~test Array~string~` might get confused if they see this PR in the release notes).

**Comment by jgreywolf** [comment]
Im not sure where my comment went - but this was by design.  any leading ~ characters are stripped before the string is even sent to this method
